resources/design-document.md

```diff
@@ -0,0 +1,335 @@
+* Save an estimate by assigning it to an existing translation client or by creating a new translation client
+* Retrieve past estimates
+* Retrieve a list of all estimates, sorted by project type
+* Retrieve a list of all estimates, sorted by translation client
+* Link an existing estimate to a new translation case
+* Update the estimate with notes on negotiation or other details (saved estimate is read-only, notes can be edited)
+* Archive (soft-delete) an estimate
+
+#### Statistics
+* Pull statistics from the translation cases
+  * Words translated per hour compared by project type
+  * Most productive time of day for each project type
+  * Largest project type by case count
+  * Total translation time since beginning of year / month
+  * Translation clients with the fastest payment turnaround time
+  * etc.
+
+
+### 4.2. Out of Scope
+
+* Linking accounts
+* Storing translated documents
+* Making translation keywords searchable
+* Creating estimates that can be sent to customers directly
+
+# 5. Proposed Architecture Overview
+
+This initial iteration will provide the minimum loveable product (MLP) including creating, retrieving, updating and 
+  archiving translation cases and translation clients, as well as sorting these objects in useful ways.
+
+I will use API Gateway and Lambda to create seven endpoints (`GetTranslationCase`, `CreateTranslationCase`, 
+`UpdateTranslationCase`, `ArchiveTranslationCase`, `GetTranslationClient`, `UpdateTranslationClient`, 
+  `ArchiveTranslationClient`) that will handle the creation, update, retrieval and archiving of translation clients 
+  and translation cases.
+
+I will store the translation clients and translation cases in tables in DynamoDB.
+
+TranslationTracker will also provide a web interface for users to manage their translation cases and translation 
+  clients. A main page providing lists of the translation cases with the soonest upcoming deadlines and translation 
+  clients with most recent activity will allow creation of new translation cases and link off to pages per-translation 
+  client and per-translation case to update translation progress and other details.
+
+# 6. API
+
+## 6.1. Public Models
+
+```
+// TranslationCaseModel
+
+String translationCaseId;
+String translationClientId;
+String sourceTextTitle;
+String sourceTextAuthor;
+String translatedTitle;
+String caseNickname;
+Double rate;
+String rateUnit;
+Integer count;
+String countUnit;
+Double grossPayment;
+Double taxRate;
+String projectType;
+String dueDate;
+String startDate;
+String endDate;
+Boolean openCase;
+Boolean casePaid;
+Boolean rushJob;
+Map<String, Map<String, String>> progressLog;
+Double totalWorkingHours;
+Double wordsPerHour;
+
+// TranslationClientModel
+
+String translationClientId;
+String translationClientName;
+String translationClientType;
+Map<String, Map<String, String>> translationCaseHistory;
+Map<String, Map<String, String>> paymentHistory;
+Map<String, String> outstandingPayments;
+```
+
```

Comment:
How do you envision the UI for updating these map attributes?

Reply:
On the UI the user sees labeled fields where they can input data:
word count ____________
start date ____________
end date ____________
start time ____________
end time ____________
notes ____________

The outer map uses the start date + end date + start time + end time as the key, the internal map uses the field labels from the website (word count, notes, etc.) as the keys, and the user input as the values.

I am now thinking that there shouldn't be a reason for these values to be updated after they are created - would it be better to make a "TranslationCaseUpdate" immutable object instead of nested maps?

Reply:
1. I removed oustandingPayments - there is a variable in TranslationCases "casePaid" that I think I could use a GSI on instead.

2. 
On the UI the user sees labeled fields where they can input payment information (translation case ID is populated by choosing a TranslationCase associated with the current TranslationClient from a dropdown menu):

translation case ____________
payment date ____________
late _____y/n____
amount ____________
taxes deducted _____y/n____
notes ____________

The outer map uses the translationCaseId as the key, the internal map uses the field labels from the website (payment date, amount, etc.) as the keys, and the user input as the values.

I am now thinking that there shouldn't be a reason for these values to be updated after they are created - would it be better to make a "TranslationCasePayment" immutable object instead of nested maps?